Connections to dnodes slow down

AdamISZ opened the pull request:
(edited from original comment since force-pushed a new set of code, though the intention is exactly the same)

Fixes https://github.com/JoinMarket-Org/joinmarket-clientserver/issues/1435.

Prior to this commit, if connections were immediately closed,
no backoff or jitter were applied to delay the next connection attempt,
resulting in too many connection attempts to directory nodes that the
client could not reach, and spam in the logs.
After this commit, we enforce that we always shut down the ClientService
on any disconnection trigger or connection failure, and only attempt to
reconnect with a manual backoff delay (4 seconds increasing by 50% for
20 attempts, plus a few seconds jitter), for directory nodes, while for
non-directory nodes we always give up on any failure.

hellodarkness commented:
I'm testing this and it's still spamming reconnect attempts every 1-3 seconds I'm afraid.

AdamISZ commented:
> 

Thanks for that! I'll try and look at it more concretely soon (i.e. test, at least in some way). More feedback or suggestions welcome of course.

AdamISZ commented:
> I'm testing this and it's still spamming reconnect attempts every 1-3 seconds I'm afraid.

I've started trying to test this, using my signet directory node. My test consisted of two things:

1. shut down the d-node and start a yg bot
2. keep the d-node up, start the yg bot, wait for it to connect to the d-node, then shut down the d-node

I tried this test on current master ( 901e1064b3f9c382112578c6694d95e9eb02eb0b ) and on this PR. In both cases I see the 'desired' behaviour, namely lines like this, with backing off delays (with twisted logging enabled):

```
2023-01-19 13:46:31-0600 [-] Scheduling retry 1 to connect <txtorcon.socks.TorSocksEndpoint object at 0x7f26975903d0> in 2.1538196157369147 seconds.
2023-01-19 13:46:31-0600 [-] Stopping factory <txtorcon.socks._TorSocksFactory object at 0x7f26975ba400>
2023-01-19 13:46:33-0600 [-] Starting factory <txtorcon.socks._TorSocksFactory object at 0x7f26975b7df0>
2023-01-19 13:46:49-0600 [-] Scheduling retry 2 to connect <txtorcon.socks.TorSocksEndpoint object at 0x7f26975903d0> in 3.1795388656977126 seconds.
2023-01-19 13:46:49-0600 [-] Stopping factory <txtorcon.socks._TorSocksFactory object at 0x7f26975b7df0>
2023-01-19 13:46:52-0600 [-] Starting factory <txtorcon.socks._TorSocksFactory object at 0x7f26975538e0>
2023-01-19 13:47:10-0600 [-] Scheduling retry 3 to connect <txtorcon.socks.TorSocksEndpoint object at 0x7f26975903d0> in 3.885909607522231 seconds.
2023-01-19 13:47:10-0600 [-] Stopping factory <txtorcon.socks._TorSocksFactory object at 0x7f26975538e0>
2023-01-19 13:47:14-0600 [-] Starting factory <txtorcon.socks._TorSocksFactory object at 0x7f2697561460>
2023-01-19 13:47:17-0600 [-] Scheduling retry 4 to connect <txtorcon.socks.TorSocksEndpoint object at 0x7f26975903d0> in 5.3838218265300535 seconds.
2023-01-19 13:47:17-0600 [-] Stopping factory <txtorcon.socks._TorSocksFactory object at 0x7f2697561460>
2023-01-19 13:47:22-0600 [-] Starting factory <txtorcon.socks._TorSocksFactory object at 0x7f2697553a00>
2023-01-19 13:47:26-0600 [-] Scheduling retry 5 to connect <txtorcon.socks.TorSocksEndpoint object at 0x7f26975903d0> in 7.670260159049834 seconds.
2023-01-19 13:47:26-0600 [-] Stopping factory <txtorcon.socks._TorSocksFactory object at 0x7f2697553a00>
2023-01-19 13:47:33-0600 [-] Starting factory <txtorcon.socks._TorSocksFactory object at 0x7f26975b7700>
2023-01-19 13:47:38-0600 [-] Scheduling retry 6 to connect <txtorcon.socks.TorSocksEndpoint object at 0x7f26975903d0> in 12.23680705011742 seconds.
```

... so I'm currently rather at a loss. I'm presuming the behaviour we're discussing is triggered by something else entirely.

My tor version: Tor version 0.4.2.7 . I take it, this is rather old and I should try the latest?

hellodarkness commented:
> ... so I'm currently rather at a loss. I'm presuming the behaviour we're discussing is triggered by something else entirely.
>
> My tor version: Tor version 0.4.2.7 . I take it, this is rather old and I should try the latest?

I was living dangerously and testing on mainnet. I'm running Tor version 4.7.13, and I think congestion control [got introduced in that version](https://blog.torproject.org/congestion-contrl-047/). Maybe that's relevant?

[Torproject.org](https://blog.torproject.org/new-stable-releases-tor-03516-04510-and-0467/) recommends upgrading from older versions to Tor 0.4.6.7 so it may be worth upgrading your node anyway....:  _"Tor 0.4.6.7 fixes several bugs from earlier versions of Tor, including one that could lead to a denial-of-service attack. Everyone running an earlier version, whether as a client, a relay, or an onion service, should upgrade to Tor 0.3.5.16, 0.4.5.10, or 0.4.6.7."_

AdamISZ commented:
> > ... so I'm currently rather at a loss. I'm presuming the behaviour we're discussing is triggered by something else entirely.
> > My tor version: Tor version 0.4.2.7 . I take it, this is rather old and I should try the latest?
> 
> I was living dangerously and testing on mainnet. I'm running Tor version 4.7.13, and I think congestion control [got introduced in that version](https://blog.torproject.org/congestion-contrl-047/). Maybe that's relevant?
> 
> [Torproject.org](https://blog.torproject.org/new-stable-releases-tor-03516-04510-and-0467/) recommends upgrading from older versions to Tor 0.4.6.7 so it may be worth upgrading your node anyway....: _"Tor 0.4.6.7 fixes several bugs from earlier versions of Tor, including one that could lead to a denial-of-service attack. Everyone running an earlier version, whether as a client, a relay, or an onion service, should upgrade to Tor 0.3.5.16, 0.4.5.10, or 0.4.6.7."_

Yes, thanks, that matches up with my thinking since writing that post. Also, I am now looking at it on mainnet too, and am a bit clearer that I should keep the question of 'what is causing the disconnects' separate from what I'm actually trying to here, which is to fix the aberrant behaviour of 'immediately try to reconnect all the time' (which I also see, specifically on mainnet).

AdamISZ commented:
See new first comment, because I've changed the code:

https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1436#issue-1538328225

That describes what the code actually does now, at least the intention. I have been testing both on mainnet (where I can observe the repeated disconnects) and also on signet (where i can easily control the directory node, switching it on and off simulating a more 'normal' network error). In both cases I see what is intended: for directory node connections it is exponentially backing off the attempts with delays, and (in case of signet) it can actually reconnect as intended.

I spent a *lot* of time trying to find a 'deeper' solution, looking for a way to distinguish between 'this was an attempt to start a connection to an onion service which somehow immediately aborted' and 'this was a preexisting connection that was working fine and has now dropped', but these attempts were unsuccessful and perhaps don't really make sense. I'm not sure what is actually going on at the Tor level other than something like "directory node is receiving tons of requests, and its Tor daemon doesn't like it and is therefore booting people immediately after they connect".

So what I've done here is I think, for now, the only thing I could realistically do. Enforce that in either of these situations:
1. We connected and then some time later (1 second or 1 hour, either way), our connection was closed by the other side, or
2. We tried to connect but failed

... then, *if* the other side is a directory node, we keep trying, with an exponential backoff, and *otherwise* we just immediately give up (this was always the logic of the code, because non-dns are going to drop connections permanently, very often, by design).

Please test and let me know if you see the problem solved. You should still see a little "log spam" for the first 30-60 seconds I guess, after that it will go down rapidly to nearly nothing.

hellodarkness commented:
Just tested this and can confirm it's backing off on the connection attempts quite nicely.

PulpCattel commented:
Tested briefly 4e30e90f7e61d35c2a9f90fc89f8148b39fae27b.
It reduces the spam in the log a lot, I also tested that it still works okay with `g3hv4uynnmynqqq2mchf3fcm3yd46kfzmcdogejuckgwknwyq5ya6iad.onion:5222`.

@AdamISZ I played with master a bit in a debugger, you probably already know this but anyway.
The `socks` error we get is `[Failure instance: Traceback (failure with no frames): <class 'txtorcon.socks.SocksError'>: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.`
Nothing really interesting happens, we trigger the connect callback and shortly after the disconnect callback. I can't see anything weird happening client side.

Random idea, could it be we are exceeding the `basic.LineReceiver.MAX_LENGTH` attribute? We set it at `40000` but maybe a big orderbook (or something else?) is bigger than that? I've not tested this from the directory node side, but if as a client I reduce that value a lot, I get the exact same `socks` error and the same disconnect pattern with `g3hv4uynnmynqqq2mchf3fcm3yd46kfzmcdogejuckgwknwyq5ya6iad.onion:5222`.
Might be nothing relevant, IDK.

AdamISZ commented:
> @AdamISZ I played with master a bit in a debugger, you probably already know this but anyway.
The socks error we get is [Failure instance: Traceback (failure with no frames): <class 'txtorcon.socks.SocksError'>: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
Nothing really interesting happens, we trigger the connect callback and shortly after the disconnect callback. I can't see anything weird happening client side.

Yeah indeed, that was a big part of what I was investigating. That's the problem: you get exactly that, both from a "normal" disconnect if a d-node goes down after some time, as well as in this pathological case that's bothering us. So I suppose it's just that the d-nodes are just disconnecting us, possibly because of too much traffic(?). Yeah as for max length I think as per comments, it could only trip us up with tx messages (full serialized transactions for broadcast). Orderbook stuff is not going over the wire in toto, it's only passed between jmdaemon and jmclient as a very big string.

AdamISZ commented:
OK, I'm planning to merge this fairly soon as a triage to the existing problem (though ofc it'll need a release also!). Perhaps early Monday, if anyone else could try it or take a look in the next day that'd be greatly appreciated.

PulpCattel commented:
> Fixes https://github.com/JoinMarket-Org/joinmarket-clientserver/issues/1435.

I think we should keep the issue open (we can rename it now that the "spam" part is mostly solved) or create a new one.

> So I suppose it's just that the d-nodes are just disconnecting us, possibly because of too much traffic(?).

So we can track this situation. The "disconnect" part is the one interesting to me.

Especially considering that the new directory is quite possibly living on borrowed time.

AdamISZ commented:
> I think we should keep the issue open (we can rename it now that the "spam" part is mostly solved) or create a new one.

Yes, sure, my only reason was that the issue specifically talked about log spam. Sure, keep it open if you like.

> So we can track this situation. The "disconnect" part is the one interesting to me.
> Especially considering that the new directory is quite possibly living on borrowed time.

Yes I think we all well understand, at this point, that there is a big problem currently unresolved; I think what none of us really understand is exactly what's causing it. If it turns out that this change does not help with the problem (in the specific sense: it *could* help, by reducing traffic to the d-nodes, but absolutely no guarantee of that), then our JM nodes are just going to fallback to IRC traffic.

PulpCattel commented on jmdaemon/jmdaemon/onionmc.py:
```diff
@@ -513,33 +515,31 @@ def connect(self) -> None:
+                           self.respond_to_connection_failure)
             self.reconnecting_service.startService()
 
+    def respond_to_connection_success(self, x):
```
Nit: Renaming `x` here would be nice. I see it's an instance of `OnionLineProtocol`, maybe `proto` or `protocol` would work better (somewhere else we call it `p` I think, which I also don't think it's great). We could also add type hinting `-> None`.

AdamISZ replied:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1436/commits/f19c0f086f4bcbd9585dda93c870c60430005549

PulpCattel commented on jmdaemon/jmdaemon/onionmc.py:
```diff
@@ -583,19 +583,23 @@ def try_to_connect(self) -> None:
-            self.connect()
-        except OnionPeerConnectionError:
-            reactor.callLater(self.delay, self.try_to_connect)
+        #try:
```
Nit: leftover?

AdamISZ replied:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1436/commits/f19c0f086f4bcbd9585dda93c870c60430005549

PulpCattel commented on jmdaemon/jmdaemon/onionmc.py:
```diff
@@ -583,19 +583,23 @@ def try_to_connect(self) -> None:
+                 f"{self.delay} seconds.")
+        # randomize by a few seconds to minimize bursty-ness locally
+        jitter = random.randint(-1, 5)
+        reactor.callLater(self.delay + jitter, self.connect)
```
Question: Should we reset `self.delay` at some point? I suppose this ties into:

> I spent a lot of time trying to find a 'deeper' solution, looking for a way to distinguish between 'this was an attempt to start a connection to an onion service which somehow immediately aborted' and 'this was a preexisting connection that was working fine and has now dropped'

Would this mean that every time the directory disconnects, say once every few days because of natural/innocous causes, we would keep expanding the delay and eventually give up entirely?

AdamISZ replied:
> Would this mean that every time the directory disconnects, say once every few days because of natural/innocous causes, we would keep expanding the delay and eventually give up entirely?

The delay is currently capped at 10k seconds, so a bit less than 3 hours.

It may well be possible to track 'how long have we been connected', and if that is a reasonably long interval, then reset `self.delay` back to 4s (or whatever). I'll just leave it for now though (if it ever got up to 10k seconds, then something was quite badly wrong at some point in the past, during the running of this bot instance ... until we're really sure we know what this problem is, and how to handle it, I don't want to be too opinionated or sophisticated .. the more logic I add the more chance something is wrong).

AdamISZ replied:
Doh, what am I saying, it's not *capped* at 10k seconds, it just gives up at that point. I'm not sure why I did that, I will change it.

AdamISZ replied:
Fixed that in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1436/commits/79eaf4ab99c0d18d5a66f2dc773aaca45654cec9 - it should now be as discussed above.

PulpCattel commented on jmdaemon/jmdaemon/onionmc.py:
```diff
@@ -583,19 +583,23 @@ def try_to_connect(self) -> None:
-        except OnionPeerConnectionError:
-            reactor.callLater(self.delay, self.try_to_connect)
+        #try:
+        log.info(f"Going to reattempt connection to {self.peer_location()} in "
+                 f"{self.delay} seconds.")
```
```suggestion
        log.info(f"Going to reattempt connection to {self.peer_location()} in "
                 f"{self.delay + jitter} seconds.")
```
Should we add the `jitter` here? We could create a variable `real_delay` or something.

AdamISZ replied:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1436/commits/f19c0f086f4bcbd9585dda93c870c60430005549

PulpCattel reviewed (commented):
Tested again a little bit. On signet I also tested a small CoinJoin and it correctly connects to both directories and makers.

Left some nits and a question.

AdamISZ commented:
Squashed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1436/commits/9a412d8be04b6f87199951686319c6b0dd55fed1 ready to merge.

AdamISZ commented:
Sorry if this is a little presumptuous, but as per earlier comment I am now going to merge without further delay. We need to include this fix in a release, and soon.

There is no assertion that the problem (most likely, related to how our d-nodes handle large amounts of traffic) is fully solved, but the current situation is not tenable. I'd urge anyone paying attention to run bots with latest master after this merge, and keep an eye on logs.
